async_simple/Signal.h

```diff
@@ -0,0 +1,453 @@
+
+#ifndef ASYNC_SIMPLE_USE_MODULES
+
+#include <assert.h>
+#include <any>
+#include <atomic>
+#include <bit>
+#include <cstdint>
+#include <memory>
+#include <system_error>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+
+#include "async_simple/Common.h"
+#include "util/move_only_function.h"
+#endif  // ASYNC_SIMPLE_USE_MODULES
+
+namespace async_simple {
+
+enum SignalType : uint64_t {
+    none = 0,
+    // low 32 bit signal only trigger once.
+    // regist those signal's handler after triggered will failed and return
+    // false.
+    terminate = 1,  // signal type: terminate
+    // 1-16 bits reserve for async-simple
+    // 17-32 bits could used for user-defined signal
+
+    // high 32 bit signal, could trigger multiple times, emplace an handler
+    // 33-48 bits reserve for async-simple
+    // 49-64 bits could used for user-defined signal
+    // after signal triggered will always success and return true.
+    all = UINT64_MAX,
+};
+
+class Slot;
+
+class Signal : public std::enable_shared_from_this<Signal> {
+private:
+    using Handler = util::move_only_function<void(SignalType, Signal*)>;
+    using TransferHandler = util::move_only_function<void(SignalType)>;
+    struct HandlerManager {
+        uint8_t _index;
+        std::atomic<Handler*> _handler;
+        HandlerManager* _next;
+        HandlerManager(uint8_t index, HandlerManager* next)
+            : _index(index), _next(next) {}
+        void releaseNodes() {
+            for (HandlerManager* tmp = _next; tmp != nullptr; tmp = _next) {
+                _next = _next->_next;
+                delete tmp;
+            }
+            delete this;
+        }
+        ~HandlerManager();
+        static inline Handler emittedTag;
+    };
+    struct PrivateConstructTag {};
+    struct Node {
+        Node(SignalType filter)
+            : _handlerManager(), _filter(filter), _next(nullptr) {}
+        ~Node();
+        void releaseNodes() {
+            for (Node *node = this->_next, *next_node; node != nullptr;
+                 node = next_node) {
+                next_node = node->_next;
+                delete node;
+            }
+            delete this;
+        }
+        bool triggered(SignalType state) const noexcept {
+            return static_cast<uint64_t>(state) &
+                   static_cast<uint64_t>(
+                       _filter.load(std::memory_order_acquire));
+        }
+        void setFilter(SignalType filter) noexcept {
+            _filter.store(filter, std::memory_order_release);
+        }
+        SignalType getFilter() const noexcept {
+            return _filter.load(std::memory_order_acquire);
+        }
+        void invoke(SignalType type, SignalType signal, Signal* self,
+                    std::atomic<Handler*>& handlerPtr);
+        void operator()(SignalType type, Signal* self);
+        std::atomic<HandlerManager*> _handlerManager;
+        std::atomic<TransferHandler*> _transferHandler;
```

Comment:
Is it really necessary to make it atomic?

Reply:
Yes. The thread which call signal->emit() and the thread which call slot->emplace() will read/write the node concurrently. So it should be atomic.